webhook.go

```diff
@@ -0,0 +1,224 @@
+	Created   time.Time `json:"created"` // The date and time the event was created. "2018-10-03T04:58:35.385Z".
+	AccountID string    `json:"account_id"`
+	AppID     string    `json:"app_id"`
+	PaymentID string    `json:"payment_id"`
+}
+
+func (CallbackCommon) isCallback() {}
+
+type PaymentCallback struct {
+	CallbackCommon
+	Data Payment `json:"data"`
+}
+
+type AuthorizationCallback struct {
+	CallbackCommon
+	Data Authorization `json:"data"`
+}
+
+type CaptureCallback struct {
+	CallbackCommon
+	Data Capture `json:"data"`
+}
+
+type VoidCallback struct {
+	CallbackCommon
+	Data Void `json:"data"`
+}
+
+type RefundCallback struct {
+	CallbackCommon
+	Data Refund `json:"data"`
+}
+
+type privateKeyProvider interface {
+	// PrivateKey should return private key for given business unit. It is used to validate request signature.
+	// For unknown business units (including empty appID) it should return (nil, nil).
+	PrivateKey(ctx context.Context, appID string) ([]byte, error)
+}
+
+type PrivateKeyProviderFunc func(ctx context.Context, appID string) ([]byte, error)
+
+func (f PrivateKeyProviderFunc) PrivateKey(ctx context.Context, appID string) ([]byte, error) {
+	return f(ctx, appID)
+}
+
+type ErrBadRequest struct {
+	Err error
+}
+
+func (e ErrBadRequest) Error() string { return e.Err.Error() }
+
+func (e ErrBadRequest) Unwrap() error { return e.Err }
+
+// DecodeWebhookRequest decodes PaymentsOS webhook http request into callback entity.
+// Supports webhook version >= 1.2.0
+//
+// Will return ErrBadRequest if the error is permanent and request should not be retried.
+// Bear in mind that your webhook handler should respond with 2xx status code anyway, otherwise PaymentsOS will continue
+// resending this request.
+// ErrBadRequest errors include:
+// 	* wrong body format (broken/invalid json, ...)
+//  * validation error (missing required fields or headers, unexpected values)
+// 	* unknown business unit (app_id)
+//  * incorrect request signature
+func DecodeWebhookRequest(ctx context.Context, r *http.Request, keyProvider privateKeyProvider) (Callback, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "read request body")
+	}
+
+	// Validate request signature
+	signature, err := CalculateWebhookSignature(ctx, body, r.Header, keyProvider)
+	if err != nil {
+		return nil, errors.WithMessage(err, "calculate request signature")
+	}
+	if "sig1="+signature != r.Header.Get("signature") {
+		return nil, ErrBadRequest{errors.New("incorrect signature")}
+	}
+
+	// Decode request into appropriate entity type based on "event-type" header
+	eventType := r.Header.Get(headerEventType)
+	common := CallbackCommon{
+		EventType:      eventType,
+		XPaymentsOSEnv: r.Header.Get("x-payments-os-env"),
+		XZoozRequestID: r.Header.Get("x-zooz-request-id"),
+	}
+	var cbRef Callback
+	switch true {
+	case strings.HasPrefix(eventType, "payment.payment."):
+		cbRef = &PaymentCallback{CallbackCommon: common}
```

Comment:
```suggestion
		cbRef = &PaymentCallback{CallbackCommon: common} // it is not possible to Unmarshal without pointer
```

Reply:
However I prefer not to use pointers
```suggestion
		cbRef = PaymentCallback{CallbackCommon: common}
```

To make it possible we can add a little bit code duplication:
```go
type Callback interface {
	unmarshal(data []byte) (Callback, error)
}

type PaymentCallback struct {
	XXX int `json:"xxx"`
}

type AuthorizationCallback struct {
	YYY int `json:"yyy"`
}

func (c PaymentCallback) unmarshal(data []byte) (Callback, error) {
	err := json.Unmarshal(data, &c)
	return c, err
}

func (c AuthorizationCallback) unmarshal(data []byte) (Callback, error) {
	err := json.Unmarshal(data, &c)
	return c, err
}

func TestXXX(t *testing.T) {
	var a Callback
	const data = `{"xxx": 123, "yyy": 456}`
	a = PaymentCallback{}
	a, err := a.unmarshal([]byte(data))
	require.NoError(t, err)
	require.Equal(t, PaymentCallback{XXX: 123}, a)
	a = AuthorizationCallback{}
	a, err = a.unmarshal([]byte(data))
	require.NoError(t, err)
	require.Equal(t, AuthorizationCallback{YYY: 456}, a)
}
```

Reply:
Pointers are used only internally for unmarshalling.  Before returning, entities are dereferenced back into values.

Reply:
> Pointers are used only internally for unmarshalling. Before returning, entities are dereferenced back into values.

I know, that's why I suggested to add comment.

> It is not obvious that entity, returned by unmarshal() will be unmarshalled over a copy of original entity (to keep header values).

All unmarshallers I saw so far are able to work with pre-filled structs.

Reply:
When unmarshalling is done into the same struct on which it is called - yes.

Here you suggest to do a (shallow) copy and then unmarshall over it. Potentially modifying original struct as well if there were pointers.